src/passes/StackIR.cpp

```diff
@@ -0,0 +1,393 @@
+    // this on the Binaryen IR, so we are assuming that no previous opt
+    // has changed the interaction of local operations.
+    // TODO: we can do this a lot faster, as we just care about linear
+    //       control flow.
+    LocalGraph localGraph(func);
+    localGraph.computeInfluences();
+    // We maintain a stack of relevant values. This contains:
+    //  * a null for each actual value that the value stack would have
+    //  * an index of each SetLocal that *could* be on the value
+    //    stack at that location.
+    const Index null = -1;
+    std::vector<Index> values;
+    // We also maintain a stack of values vectors for control flow,
+    // saving the stack as we enter and restoring it when we exit.
+    std::vector<std::vector<Index>> savedValues;
+#ifdef STACK_OPT_DEBUG
+    std::cout << "func: " << func->name << '\n' << insts << '\n';
+#endif
+    for (Index i = 0; i < insts.size(); i++) {
+      auto* inst = insts[i];
+      if (!inst) continue;
+      // First, consume values from the stack as required.
+      auto consumed = getNumConsumedValues(inst);
+#ifdef STACK_OPT_DEBUG
+      std::cout << "  " << i << " : " << *inst << ", " << values.size() << " on stack, will consume " << consumed << "\n    ";
+      for (auto s : values) std::cout << s << ' ';
+      std::cout << '\n';
+#endif
+      // TODO: currently we run dce before this, but if we didn't, we'd need
+      //       to handle unreachable code here - it's ok to pop multiple values
+      //       there even if the stack is at size 0.
+      while (consumed > 0) {
+        assert(values.size() > 0);
+        // Whenever we hit a possible stack value, kill it - it would
+        // be consumed here, so we can never optimize to it.
+        while (values.back() != null) {
+          values.pop_back();
+          assert(values.size() > 0);
+        }
+        // Finally, consume the actual value that is consumed here.
+        values.pop_back();
+        consumed--;
+      }
+      // After consuming, we can see what to do with this. First, handle
+      // control flow.
+      if (isControlFlowBegin(inst)) {
+        // Save the stack for when we end this control flow.
+        savedValues.push_back(values); // TODO: optimize copies
+        values.clear();
+      } else if (isControlFlowEnd(inst)) {
+        assert(!savedValues.empty());
+        values = savedValues.back();
+        savedValues.pop_back();
+      } else if (isControlFlow(inst)) {
+        // Otherwise, in the middle of control flow, just clear it
+        values.clear();
+      }
+      // This is something we should handle, look into it.
+      if (isConcreteType(inst->type)) {
+        bool optimized = false;
+        if (auto* get = inst->origin->dynCast<GetLocal>()) {
+          // This is a potential optimization opportunity! See if we
+          // can reach the set.
+          if (values.size() > 0) {
+            Index j = values.size() - 1;
+            while (1) {
```

Comment:
Is this any less weird?
```
for (Index j1 = values.size(); j1 != 0; --j1) {
  Index j = j1 - 1;
```
The `while (1)` is kinda weird... but so is the off-by-one for loop :/

Reply:
Yeah, it's weird either way I think... it's just awkward to iterate down to 0 on an unsigned number.